manifests/components/kibana.jsonnet

```diff
@@ -38,18 +39,69 @@ local strip_trailing_slash(s) = (
     },
   },
 
+  // List of Kibana plug-ins to install
+  plugins:: {
+    /* Example:
+    logtrail: {
+      version: "v0.1.31",
+      url: "https://github.com/sivasamyk/logtrail/releases/download/v0.1.31/logtrail-6.7.1-0.1.31.zip",
+    },
+    */
+  },
+
   es: error "elasticsearch is required",
 
   serviceAccount: kube.ServiceAccount($.p + "kibana") + $.metadata {
   },
 
+  pvc: kube.PersistentVolumeClaim($.p + "kibana-plugins") + $.metadata {
+    storage: "2Gi",
+  },
+
   deploy: kube.Deployment($.p + "kibana") + $.metadata {
     spec+: {
       template+: {
         spec+: {
           securityContext: {
             fsGroup: 1001,
           },
+          volumes_+: {
+            plugins: kube.PersistentVolumeClaimVolume($.pvc),
+          },
+          initContainers_+: {
+            kibana_plugins_install: kube.Container("kibana-plugins-install") {
+              image: KIBANA_IMAGE,
+              securityContext: {
+                allowPrivilegeEscalation: false,
+              },
+              env_+: {
+                local plugins = [{name: name} + $.plugins[name] for name in std.objectFields($.plugins)],
+                KIBANA_PLUGINS: std.join('\n', ["%(name)s,%(version)s,%(url)s" % plugin for plugin in plugins]),
+              },
+              command: [
+                "/bin/bash",
+                "-c",
+                |||
+                  set -e
+                  set -x
+                  exec &> >(tee /opt/bitnami/kibana/plugins/log)
+                  rm -rf /opt/bitnami/kibana/plugins/lost+found
+                  for i in ${KIBANA_PLUGINS}
+                  do
+                    IFS=',' read -ra PLUGIN <<< "$i"
+                    /opt/bitnami/kibana/bin/kibana-plugin remove --silent "${PLUGIN[0]}" || :
```

Comment:
If I'm reading this correctly, I think it is _always_ removing and reinstalling each plugin on every pod restart.  If we're going to do that, then we may as well not use a PVC?  Is there some way we can skip this if the plugin/version is already installed?

Reply:
I note the helm chart does (basically) the following to check if a plugin is already installed:
```sh
kibana-plugin list | grep -q $name@$version
```

Unlike the helm script, we want to just always make the installed plugin state equal to the user settings - the helm chart has a mix of "install/upgrade" and "reset".  I think I would do something like (untested):

```jsonnet
local wanted = std.join("\n",
  ["%s@%s,%s" % [k, $.plugins[k].version, $.plugins[k].url] for k in std.objectFields($.plugins)]);
|||
  set -e
  cd /tmp
  echo %s | sort > wanted.list 
  kibana-list | grep @  | sort > installed.list
  join -v2 -t, -j1 wanted.list installed.list |
  while read plugin; do
    kibana-plugin remove ${plugin%@*}
  done
  join -v1 -t, -j1 -o1.2 wanted.list installed.list |
  while read url; do
    kibana-plugin install --no-optimize $url
  done
||| % std.escapeStringBash(wanted)
```

(`join` is one of the most underappreciated shell commands :)

Reply:
Nice. I've implemented your very nice suggestion and also tested it.